test-tests/shared/src/test/scala/zio/test/environment/ClockSpec.scala

```diff
@@ -0,0 +1,157 @@
+package zio.test.environment
+
+import java.time.ZoneId
+import java.util.concurrent.TimeUnit
+
+import zio._
+import zio.clock._
+import zio.duration.Duration._
+import zio.duration._
+import zio.test.Assertion._
+import zio.test.TestAspect._
+import zio.test._
+import zio.test.environment.TestClock._
+
+object ClockSpec
+    extends ZIOBaseSpec(
+      suite("ClockSpec")(
+        testM("sleep does not require passage of clock time") {
+          for {
+            latch <- Promise.make[Nothing, Unit]
+            _     <- (sleep(10.hours) *> latch.succeed(())).fork
+            _     <- adjust(11.hours)
+            _     <- latch.await
+          } yield assert((), anything)
+        } @@ nonFlaky(100),
+        testM("sleep delays effect until time is adjusted") {
+          for {
+            ref    <- Ref.make(true)
+            _      <- (sleep(10.hours) *> ref.set(false)).fork
+            _      <- adjust(9.hours)
+            result <- ref.get
+          } yield assert(result, isTrue)
+        } @@ nonFlaky(100),
+        //todo  remove `@@ after(setTime(0.hours))` if possible. This doesn't depend on absolute time so I don't think
+        // it should rely on resetting the clock. Fails intermittently without it.
+        testM("sleep correctly handles multiple sleeps") {
+          for {
+            latch1 <- Promise.make[Nothing, Unit]
+            latch2 <- Promise.make[Nothing, Unit]
+            ref    <- Ref.make("")
+            _      <- (sleep(3.hours) *> ref.update(_ + "World!") *> latch2.succeed(())).fork
+            _      <- (sleep(1.hours) *> ref.update(_ + "Hello, ") *> latch1.succeed(())).fork
+            _      <- adjust(2.hours)
+            _      <- latch1.await
+            _      <- adjust(2.hours)
+            _      <- latch2.await
+            result <- ref.get
+          } yield assert(result, equalTo("Hello, World!"))
+        } @@ after(setTime(0.hours))
+          @@ nonFlaky(100),
+        testM("sleep correctly handles new set time") {
+          for {
+            latch <- Promise.make[Nothing, Unit]
+            _     <- (sleep(10.hours) *> latch.succeed(())).fork
+            _     <- setTime(11.hours)
+            _     <- latch.await
+          } yield assert((), anything)
+        } @@ nonFlaky(100)
+          @@ timeout(10.seconds),
+        //todo fix - times out
+        testM("sleep does sleep instantly when sleep duration less than set time") {
```

Comment:
The problem here is that since the `sleep(10.hours)` call is occurring on the main fiber the fiber time is being incremented, so in the second run when you sleep for another 10 hours the fiber is not going to wake up until the clock time is 20 hours, but you are only setting the time to 11 hours. You can fix this by changing `setTime(11.hours)` to `adjust(10.hours)` so the clock time stays in sync with the fiber time.

Reply:
that'll make it run, but we wouldn't be testing the same thing anymore.
spec is "sleep does sleep instantly when sleep duration _less_ than _set time_"
but the changes you've proposed would remove set time all together and the sleep duration would be identical rather than less than.

I think there's a bigger issue here with `nonFlaky` - is there a way to make it sync the clock with fiberTime at the end of each iteration? or provide a new environment? or at least the option for one? At the moment is has the rather ironic consequence of making otherwise solid tests flaky unless you dance the dance of sleep and adjust perfectly every time!

Reply:
Maybe we change the test label to "sleep does sleep instantly when sleep duration less than or equal to clock time"? The idea of the test is more about making sure effects run than the specific `setTime` versus `adjust` method and making sure tests effects run if the time is less than or equal is actually a stronger guarantee that we want. There are other tricks we could do to keep the clock time and fiber time in sync if you want but this seemed like the simplest way to do.

I had the same thought about `nonFlaky`. Unfortunately I don't think there is a general solution. The problem is that from the perspective of `nonFlaky` `R` is completely polymorphic. It is easy to think in a case like this why can't we just provide a new copy of the environment. But the environment might be something like a database connection where we can't just create a new database connection. We could create a new `TestClock` in each test here and then be sure to `provide` it to each effect but that doesn't solve the problem in general.